heirloom/echo/echo.1
.\"
.\" Sccsid @(#)echo.1	1.3 (gritter) 6/30/05
.\" Derived from echo(1), Unix 7th Edition:
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"   Redistributions of source code and documentation must retain the
.\"    above copyright notice, this list of conditions and the following
.\"    disclaimer.
.\"   Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"   All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed or owned by Caldera
.\"      International, Inc.
.\"   Neither the name of Caldera International, Inc. nor the names of
.\"    other contributors may be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" USE OF THE SOFTWARE PROVIDED FOR UNDER THIS LICENSE BY CALDERA
.\" INTERNATIONAL, INC. AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL CALDERA INTERNATIONAL, INC. BE
.\" LIABLE FOR ANY DIRECT, INDIRECT INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
.\" BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
.\" OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
.\" EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.TH ECHO 1 "6/30/05" "Heirloom Toolchest" "User Commands"
.SH NAME
echo \- echo arguments
.SH SYNOPSIS
\fBecho\fR [\fIargument \fR...]
.SH DESCRIPTION
.I Echo
writes its
.I arguments
separated by blanks and terminated by
a newline on the standard output.
.PP
.I Echo
is useful for producing diagnostics in
shell programs and for writing constant data on pipes.
To send diagnostics to the standard error file, do
`echo ... 1>&2'.
.PP
The following escapes sequences are recognized in the
.IR arguments :
.TP 7n
.B \ea
Prints an alert character
.RB ( /usr/5bin/posix/echo
and
.B /usr/5bin/posix2001/echo
only).
.TP
.B \eb
Prints a backspace character.
.TP
.B \ec
Causes the command to return immediately.
Any following characters are ignored,
and the terminating newline is not printed.
.TP
.B \ef
Prints a formfeed character.
.TP
.B \en
Prints a newline character.
.TP
.B \er
Prints a carriage-return character.
.TP
.B \et
Prints a tabulator character.
.TP
.B \ev
Prints a vertical tabulator character.
.TP
.B \e\e
Prints a backslash character.
.TP
.BI \e0 nnn
Prints the character (byte) with octal value
.IR nnn .
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale (7).
.TP
.B LC_CTYPE
Determines the scope of backslash as an escape character.
.TP
.B PATH
If
.I /usr/ucb
precedes
.I /usr/5bin
or
.I /usr/bin
in the setting of this
variable and the first argument is
.BR \-n ,
the terminating newline is not printed,
and no escape sequences are recognized.
.TP
.B SYSV3
If this variable is set,
the
.I \-n
argument is also interpreted,
but escape sequences are processed as usual.
.SH "SEE ALSO"
printf(1)
.SH NOTES
It is recommended for portability that
.IR printf (1)
is used instead of
.I echo
if backslashes are to be printed
or if the trailing newline is to be omitted.


heirloom-sh/echo.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/


/*	Portions Copyright (c) 1988, Sun Microsystems, Inc.	*/
/*	All Rights Reserved.					*/

/*
 * Portions Copyright (c) 2005 Gunnar Ritter, Freiburg i. Br., Germany
 *
 * Sccsid @(#)echo.c	1.8 (gritter) 6/30/05
 */
/* from OpenSolaris "echo.c	1.16	05/06/08 SMI" */

/*
 *	UNIX shell
 */
#include	"defs.h"

#define	_iBCS2

#define	exit(a)	flushb(); return (a)

int
echo(int argc, unsigned char **argv)
{
	register unsigned char	*cp;
	register int	i, wd;
	int	nflg = 0;
	int	j;
	int	len;
	wchar_t	wc;

#ifdef	_iBCS2		/* SCO compatibility support */
	struct namnod   *sysv3;
	int	do_sysv3 = 0;

	sysv3 = findnam("SYSV3");
	if (sysv3 && (sysv3->namflg & (N_EXPORT | N_ENVNAM)))
		do_sysv3 = 1;

	/* Do the -n parsing if sysv3 is set or if ucb_builtsin is set */
	if (ucb_builtins && !do_sysv3) {
#else
	if (ucb_builtins) {
#endif /* _iBCS2 */

		nflg = 0;
		if (argc > 1 && argv[1][0] == '-' && argv[1][1] == 'n') {
			nflg++;
			argc--;
			argv++;
		}

		for (i = 1; i < argc; i++) {
			sigchk();

			for (cp = argv[i]; *cp; cp++) {
				prc_buff(*cp);
			}

			if (i < argc-1)
				prc_buff(' ');
		}

		if (nflg == 0)
			prc_buff('\n');
		exit(0);
	} else {
		if (--argc == 0) {
			prc_buff('\n');
			exit(0);
		}
#ifdef  _iBCS2
		if (do_sysv3) {
			if (argc > 1 && argv[1][0] == '-' &&
					argv[1][1] == 'n') {
				nflg++;
				/* Step past the -n */
				argc--;
				argv++;
			}
		}
#endif /* _iBCS2 */

		for (i = 1; i <= argc; i++)
		{
			sigchk();
			for (cp = argv[i]; *cp; cp++) {
				if ((len = nextc(&wc, (char *)cp)) <= 0) {
					prc_buff(*cp);
					continue;
				}

				if (wc == '\\') {
					switch (*++cp) {
#ifdef	SUS
					case 'a':
						prc_buff('\a');
						continue;
#endif	/* SUS */
					case 'b':
						prc_buff('\b');
						continue;
					case 'c':
						exit(0);

					case 'f':
						prc_buff('\f');
						continue;

					case 'n':
						prc_buff('\n');
						continue;

					case 'r':
						prc_buff('\r');
						continue;

					case 't':
						prc_buff('\t');
						continue;

					case 'v':
						prc_buff('\v');
						continue;

					case '\\':
						prc_buff('\\');
						continue;
					case '0':
						j = wd = 0;
						while ((*++cp >= '0' &&
						*cp <= '7') && j++ < 3) {
							wd <<= 3;
							wd |= (*cp - '0');
						}
						prc_buff(wd);
						--cp;
						continue;

					default:
						cp--;
					}
					prc_buff(*cp);
					continue;
				} else {
					for (; len > 0; len--)
						prc_buff(*cp++);
					cp--;
					continue;
				}
			}
#ifdef	_iBCS2
			/* Don't do if don't want newlines & out of args */
			if (!(nflg && i == argc))
#endif /* _iBCS2 */
				prc_buff(i == argc? '\n': ' ');
		}
		exit(0);
	}
}
